### PR TITLE
Lower TransposeOp to a view op if the order of the dimensions whose value is not 1 does not change

### PR DESCRIPTION
**Pull request by tungld**
This patch adds a special case when lowering ONNXTransposeOp where if the order of the dimensions whose value is not 1 does not change, then it is safe to lower ONNXTransposeOp to a view op.

Signed-off-by: Tung D. Le <tung@jp.ibm.com>

### CODE REVIEW

File: src/Dialect/ONNX/Rewrite.cpp [hunk 1]
```diff
@@ -63,6 +63,36 @@ Type getReturnTypeForMatMulOpND2D(Value A, Value B) {
...
+  return (llvm::count_if(dims, [](int64_t d) { return d == -1; }) <= 1);
+}
+
+// Check that permutation does not change the order of dimensions except those
+// of size 1.
+bool hasNoOrderChangeExceptDimOne(Value v, ArrayAttr permAttr) {
+  if (!onnx_mlir::hasShapeAndRank(v))
+    return false;
+  ArrayRef<int64_t> dims = v.getType().cast<ShapedType>().getShape();
+  SmallVector<int64_t> originalAxes;
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Dialect/ONNX/Rewrite.cpp]: Done. Thanks!

File: src/Dialect/ONNX/Rewrite.cpp [hunk 2]
```diff
@@ -63,6 +63,36 @@ Type getReturnTypeForMatMulOpND2D(Value A, Value B) {
...
+  for (uint64_t axis = 0; axis < dims.size(); ++axis)
+    if (dims[axis] != 1)
+      originalAxes.emplace_back(axis);
+
+  SmallVector<int64_t> permutedAxes;
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Dialect/ONNX/Rewrite.cpp]: Done. Thanks!

File: src/Dialect/ONNX/Rewrite.cpp [hunk 3]
```diff
@@ -63,6 +63,36 @@ Type getReturnTypeForMatMulOpND2D(Value A, Value B) {
...
+  for (Attribute val : permAttr.getValue()) {
+    IntegerAttr attr = val.cast<IntegerAttr>();
+    assert(attr && "Element in ArrayAttr is not IntegerAttr");
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Dialect/ONNX/Rewrite.cpp]: Done. Thanks!

File: test/mlir/onnx/onnx_canonicalization.mlir [hunk 4]
```diff
@@ -169,6 +169,18 @@ func @test_transpose_removal(%arg0: tensor<10x11x12x13xf32>) -> tensor<10x11x12x
 
 // -----
 
+// Check the replacement of transpose by reshape because of the dimension order unchanged.
```
[the remark this answers is not in the document]
Comment (tungld) [reply test/mlir/onnx/onnx_canonicalization.mlir]: We do not count the dimensions with size 1. I updated the comment to make it clear. Thanks!

**Comment by tungld** [comment]
@etiotto I move the code into the lowering code of ONNXTransposeOp, which is more appropriate, e.g. we can handle unknown dimensions.

**Comment by etiotto** [comment]
@tungld thanks, makes sense to me now!

**Review by etiotto — APPROVED** [review]
LGTM

**Pull request merged**